ios/engine/KMEI/KeymanEngine/Classes/KeyboardMenuView.swift

```diff
@@ -0,0 +1,356 @@
+        viewPosX = keyFrame.origin.x
+      } else {
+        adjX = -viewPosX
+        viewPosX = 0
+      }
+    } else if (viewPosX + viewWidth) > screenWidth {
+      if ((keyFrame.origin.x + keyFrame.size.width) + borderRadius * 1.0) > screenWidth {
+        adjX = viewPosX - keyFrame.origin.x
+        viewPosX = keyFrame.origin.x + 2 * adjX
+      } else {
+        adjX = (screenWidth - viewWidth) - viewPosX
+        viewPosX = adjX + viewPosX
+      }
+    }
+
+    menuFrame = CGRect(x: viewPosX, y: viewPosY, width: viewWidth, height: viewHeight)
+
+    backgroundColor = UIColor(white: 0.0, alpha: 0.25)
+    tag = 1
+    isUserInteractionEnabled = true
+    let menuView = UIView(frame: menuFrame)
+    menuView.tag = 2
+    menuView.backgroundColor = UIColor.clear
+    tableView = UITableView(frame: containerFrame)
+    tableView?.delegate = self
+    tableView?.dataSource = self
+    tableView?.separatorStyle = .none
+    tableView?.backgroundColor = UIColor.clear
+    tableView?.layer.cornerRadius = (borderRadius * 1.0 - strokeWidth)
+    menuView.addSubview(tableView!)
+    addSubview(menuView)
+    let tapGesture = UITapGestureRecognizer(target: self, action: #selector(self.tapAction))
+    tapGesture.delegate = self
+    tapGesture.numberOfTapsRequired = 1
+    tapGesture.numberOfTouchesRequired = 1
+    addGestureRecognizer(tapGesture)
+
+  }
+
+  public required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  public override func draw(_ rect: CGRect) {
+    let keyWidth = keyFrame.width
+    let keyHeight = keyFrame.height
+    let viewWidth = menuFrame.width
+    let viewHeight = menuFrame.height
+
+    guard let context = UIGraphicsGetCurrentContext() else {
+      return
+    }
+    UIGraphicsPushContext(context)
+    context.setLineJoin(.round)
+    context.setLineWidth(strokeWidth * 2)
+    context.setStrokeColor(borderColor.cgColor)
+    context.setFillColor(bgColor.cgColor)
+    var keyLeft = menuFrame.origin.x + (viewWidth - keyWidth) / 2.0 + strokeWidth - adjX
+    var keyRight  = menuFrame.origin.x + (viewWidth - keyWidth) / 2.0 + keyWidth - strokeWidth - adjX
+    keyLeft = (keyLeft * 10).rounded(.up) / 10
+    keyRight = (keyRight * 10).rounded(.down) / 10
+
+    let midY = menuFrame.origin.y + viewHeight - keyHeight
+    let viewLeft = menuFrame.origin.x + strokeWidth
+    let viewRight = menuFrame.origin.x + viewWidth - strokeWidth
+    let viewTop = menuFrame.origin.y + strokeWidth
+    let viewBottom = menuFrame.origin.y + viewHeight - strokeWidth * 1.25
+    let viewMid = menuFrame.origin.x + viewWidth / 2.0
+    let r = borderRadius - strokeWidth
+    let r2 = borderRadius * 1.0 - strokeWidth
+    let r3 = borderRadius * 1.0 - strokeWidth
+
+    context.beginPath()
+    context.move(to: CGPoint(x: viewMid, y: viewTop))
+    context.addArc(tangent1End: CGPoint(x: viewLeft, y: viewTop),
+                   tangent2End: CGPoint(x: viewLeft, y: midY), radius: r2)
```

Comment:
This looks a lot like [this block from KeyPreviewView.swift](https://github.com/keymanapp/keyman/pull/378/files#diff-f5aa2863ba68d988ee3a9a348ef55cb3R115) - possibly a good opportunity to refactor?

I mean, both files execute nearly the same pattern twice!